File: scripts/txline_proof.ts
import * as anchor from "@coral-xyz/anchor";

const TXLINE_API_BASE = "https://txline-dev.txodds.com/api";

function toBytes32(value: string | number[] | Uint8Array): number[] {
  if (typeof value === "string") {
    const raw = value.startsWith("0x") ? value.slice(2) : value;
    if (/^[0-9a-fA-F]{64}$/.test(raw)) {
      return Array.from(Buffer.from(raw, "hex"));
    }
    return Array.from(Buffer.from(raw, "base64"));
  }
  return Array.from(value as any);
}

function toProofNodes(nodes: any[] = []) {
  return nodes.map((node) => ({
    hash: toBytes32(node.hash),
    isRightSibling: Boolean(node.isRightSibling ?? node.is_right_sibling ?? false),
  }));
}

async function fetchJson(url: string, init?: RequestInit): Promise<any> {
  const resp = await fetch(url, init);
  if (!resp.ok) {
    throw new Error(`${url} -> ${resp.status} ${await resp.text()}`);
  }
  return resp.json();
}

export async function buildFinalSettlementProof(
  jwt: string,
  apiToken: string,
  fixtureId: number,
  seq: number,
): Promise<{
  validationTimestamp: number;
  homeScore: number;
  awayScore: number;
  proposedOutcome: number;
  validationInput: any;
  strategy: any;
}> {
  const validation = await fetchJson(`${TXLINE_API_BASE}/scores/stat-validation?fixtureId=${fixtureId}&seq=${seq}&statKeys=1002,1003`, {
    headers: {
      authorization: `Bearer ${jwt}`,
      "x-api-token": apiToken,
    },
  });

  const updateStats = validation.summary.updateStats;
  const statsToProve = validation.statsToProve ?? [];
  const statProofs = validation.statProofs ?? [];

  if (statsToProve.length < 2) {
    throw new Error("TxLINE proof payload requires at least two stats");
  }

  const validationTimestamp = Number(updateStats.minTimestamp);
  const homeScore = Number(statsToProve[0].value ?? 0);
  const awayScore = Number(statsToProve[1].value ?? 0);

  return {
    validationTimestamp,
    homeScore,
    awayScore,
    proposedOutcome: homeScore > awayScore ? 0 : awayScore > homeScore ? 2 : 1,
    validationInput: {
      ts: new anchor.BN(validationTimestamp),
      fixtureSummary: {
        fixtureId: new anchor.BN(Number(validation.summary.fixtureId)),
        updateStats: {
          updateCount: Number(updateStats.updateCount),
          minTimestamp: new anchor.BN(Number(updateStats.minTimestamp)),
          maxTimestamp: new anchor.BN(Number(updateStats.maxTimestamp)),
        },
        eventsSubtreeRoot: toBytes32(
          validation.summary.eventStatsSubTreeRoot ?? validation.summary.eventsSubTreeRoot,
        ),
      },
      fixtureProof: toProofNodes(validation.subTreeProof),
      mainTreeProof: toProofNodes(validation.mainTreeProof),
      eventStatRoot: toBytes32(validation.eventStatRoot),
      stats: statsToProve.map((stat: any, index: number) => ({
        stat,
        statProof: toProofNodes(statProofs[index] ?? []),
      })),
    },
    strategy: {
      geometricTargets: [],
      distancePredicate: null,
      discretePredicates: [
        {
          single: {
            index: 0,
            predicate: {
              threshold: Number(statsToProve[0].value),
              comparison: { equalTo: {} },
            },
          },
        },
        {
          single: {
            index: 1,
            predicate: {
              threshold: Number(statsToProve[1].value),
              comparison: { equalTo: {} },
            },
          },
        },
      ],
    },
  };
}

File: bot/txline_proof.py
from __future__ import annotations

import base64
from typing import Any


def _to_bytes32(value: str | list[int] | bytes | bytearray | memoryview) -> list[int]:
    if isinstance(value, str):
        raw = value[2:] if value.startswith("0x") else value
        if len(raw) == 64 and all(c in "0123456789abcdefABCDEF" for c in raw):
            data = bytes.fromhex(raw)
        else:
            data = base64.b64decode(raw)
    else:
        data = bytes(value)
    if len(data) != 32:
        raise ValueError(f"Expected 32 bytes, got {len(data)}")
    return list(data)


def _to_proof_nodes(nodes: list[dict[str, Any]] | None) -> list[dict[str, Any]]:
    if not nodes:
        return []
    return [
        {
            "hash": _to_bytes32(node["hash"]),
            "isRightSibling": bool(node.get("isRightSibling", node.get("is_right_sibling", False))),
        }
        for node in nodes
    ]


def build_final_settlement_proof_bundle(validation: dict[str, Any], final_record: Any) -> dict[str, Any]:
    update_stats = validation["summary"]["updateStats"]
    stats_to_prove = validation.get("statsToProve", [])
    stat_proofs = validation.get("statProofs", [])

    if len(stats_to_prove) < 2:
        raise ValueError("TxLINE settlement proof requires at least two stats")

    target_ts = int(update_stats["minTimestamp"])
    payload = {
        "ts": target_ts,
        "fixtureSummary": {
            "fixtureId": int(validation["summary"]["fixtureId"]),
            "updateStats": {
                "updateCount": int(update_stats["updateCount"]),
                "minTimestamp": target_ts,
                "maxTimestamp": int(update_stats["maxTimestamp"]),
            },
            "eventsSubtreeRoot": _to_bytes32(
                validation["summary"].get("eventStatsSubTreeRoot", validation["summary"].get("eventsSubTreeRoot"))
            ),
        },
        "fixtureProof": _to_proof_nodes(validation.get("subTreeProof")),
        "mainTreeProof": _to_proof_nodes(validation.get("mainTreeProof")),
        "eventStatRoot": _to_bytes32(validation["eventStatRoot"]),
        "stats": [
            {
                "stat": stat,
                "statProof": _to_proof_nodes(stat_proofs[index]),
            }
            for index, stat in enumerate(stats_to_prove)
        ],
    }

    strategy = {
        "geometricTargets": [],
        "distancePredicate": None,
        "discretePredicates": [
            {
                "single": {
                    "index": 0,
                    "predicate": {
                        "threshold": int(stats_to_prove[0]["value"]),
                        "comparison": {"equalTo": {}},
                    },
                }
            },
            {
                "single": {
                    "index": 1,
                    "predicate": {
                        "threshold": int(stats_to_prove[1]["value"]),
                        "comparison": {"equalTo": {}},
                    },
                }
            },
        ],
    }

    proposed_outcome = 0
    if final_record.home_score < final_record.away_score:
        proposed_outcome = 2
    elif final_record.home_score == final_record.away_score:
        proposed_outcome = 1

    return {
        "seq": final_record.seq,
        "validation_timestamp": target_ts,
        "home_score": final_record.home_score,
        "away_score": final_record.away_score,
        "proposed_outcome": proposed_outcome,
        "validation_input": payload,
        "strategy": strategy,
    }
